docs/docs/optimizing-performance.md

```diff
@@ -0,0 +1,228 @@
+
+When a component's props or state change, React decides whether an actual DOM update is necessary by comparing the newly returned element with the previously rendered one. When they are not equal, React will update the DOM.
+
+In some cases, your component can speed all of this up by overriding the lifecycle function `shouldComponentUpdate`, which is triggered before the re-rendering process starts. The default implementation of this function returns `true`, leaving React to perform the update:
+
+```javascript
+shouldComponentUpdate(nextProps, nextState) {
+  return true;
+}
+```
+
+If you know that in some situations your component doesn't need to update, you can return `false` from `shouldComponentUpdate` instead, to skip the whole rendering process, including calling `render()` on this component and below. For example, if the only way your component ever changes is when the `props.color` or the `state.count` variable changes, you could have `shouldComponentUpdate` check that:
+
+```javascript
+class CounterButton extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {count: 1};
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    if (this.props.color !== nextProps.color) {
+      return true;
+    }
+    if (this.state.count !== nextState.count) {
+      return true;
+    }
+    return false;
+  }
+
+  render() {
+    <button
+      color={this.props.color}
+      onClick={() => this.setState(state => ({count: state.count + 1}))}>
+      Count: {this.state.count}
+    </button>
+  }
+}
+```
+
+In this code, `shouldComponentUpdate` is just checking if there is any change in `props.color` or `state.count`. If those values don't change, the component doesn't update. If your component got more complex, you could use a similar pattern of doing a "shallow comparison" between all the fields of `props` and `state` to determine if the component should update. This pattern is common enough that React provides a helper to use this logic - just inherit from `React.PureComponent`. So this code is a simpler way to achieve the same thing:
+
+```js
+class CounterButton extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    this.state = {count: 1};
+  }
+
+  render() {
+    <button
+      color={this.props.color}
+      onClick={() => this.setState(state => ({count: state.count + 1}))}>
+      Count: {this.state.count}
+    </button>
+  }
+}
+```
+
+Most of the time, you can use `React.PureComponent` instead of writing your own `shouldComponentUpdate`. It only does a shallow comparison, so you can't use it when the rendering function uses deep properties of the props or state. For example, if your render function refers to `this.props.foo.bar` or `this.state.qux.quux`, you should not use `React.PureComponent`.
```

Comment:
>If you create functions or child components in your render function, PureComponent will never short-circuit rendering, since those functions or child components will always be different. You can often avoid this by creating these once during component creation rather than on every render.

I think this is a bit confusing and should be the other way around:

>If your component accepts functions or React elements as props, `PureComponent` will never short-circuit rendering, since those functions or elements will usually be different on every render. You can often avoid this by creating them once rather than on every render. For example, if you bind class methods in the constructor, they will be the same on every render, so any time you pass them down to a `PureComponent` as props, it will be able to benefit from the optimization.